src/queryinterface.js

```diff
@@ -118,59 +139,96 @@ QueryInterface.prototype.$clearQueue = function (options) {
+			var created = !!qi.options.createdDefault;
 			if(typeof result.options.wasCreated !== 'undefined') {
 				created = !!result.options.wasCreated;
 			}
 			
-			return bluebird.resolve([result.content, created]);
+			return done(bluebird.resolve([result.content, created]));
 		}
 		if (options.includeAffectedRows) {
 			var affectedRows = [];
 			if(result.options.affectedRows instanceof Array) {
 				affectedRows = result.options.affectedRows;
 			}
 			
-			return bluebird.resolve([result.content, affectedRows]);
+			return done(bluebird.resolve([result.content, affectedRows]));
+		}
+		return done(bluebird.resolve(result.content));
+	}
+}
+
+function propagationHandler(qi, options) {
+	return function(query, done) {
+		if (!options.stopPropagation && !qi.options.stopPropagation && qi.options.parent) {
+			return done(qi.options.parent.$query(options));
+		} else {
+			return done();
 		}
-		return bluebird.resolve(result.content);
-		
-	} else if (!options.stopPropagation && !this.options.stopPropagation && this.options.parent) {
-		return this.options.parent.$query(options);
-	} else if (fallbackFn){
-		return fallbackFn();
-	} else {
-		throw new Errors.EmptyQueryQueueError();
 	}
+}
+
+function fallbackHandler(qi, options) {
+	return function(query, done) {
+		var fallbackFn = options.fallbackFn || qi.options.fallbackFn;
+		if (fallbackFn) return done(fallbackFn());
+		else return done();
+	}
+}
+
+/**
+ * This is the mock method for getting results from the `QueryInterface`. This function
+ * will get the next result in the queue and return that wrapped in a promise.
+ *
+ * @instance
+ * @param {Object} [options] Options used for this query
+ * @param {Function} [options.fallbackFn] A fallback function to run if there are no results queued
+ * @param {Boolean} [options.includeCreated] Flag indicating if a `created` value should be returned with the result for this query. Defaults to false
+ * @param {Boolean} [options.includeAffectedRows] Flag indicating if the query expects `affectedRows` in the returned result parameters. Defautls to false
+ * @param {Boolean} [options.stopPropagation] Flag indicating if result queue propagation should be stopped on this query. Defaults to false
+ * @param {String} [options.method] Name of the original query: "findOne", "findOrCreate", "upsert", etc.
+ * @param {String} [options.options] Options passed to the original query method
+ * @return {Promise} resolved or rejected promise from the next item in the review queue
+ **/
+QueryInterface.prototype.$query = function (options) {
+	options = options || {};
+
+	var handlers = this._handlers.concat(
+		resultsQueueHandler(this, options),
+		propagationHandler(this, options),
+		fallbackHandler(this, options),
+		function() {
+			throw new Errors.EmptyQueryQueueError();
+		}
+	)
+	
+	// Can't use promises to chain the handlers because they will convert any error thrown by the handlers to a rejected promise.
```

Comment:
Just trying to think through this; What is the expected if the handler throws an error? Isn't a rejected promise there acceptable? Otherwise the error might end up as an uncaught exception, right?